<html><header><title>DataPortal Exception gets &quot;swallowed&quot; by OnDeserialized()</title></header><body><div style='float:right'><a href='http://cslanet.com'><img src='https://github.com/MarimerLLC/csla/raw/master/Support/Logos/csla%20win8_compact_s.png'/></a></div><p><h1>DataPortal Exception gets &quot;swallowed&quot; by OnDeserialized()</h1><span style='font-size:small'>Old forum URL: forums.lhotka.net/forums/t/4714.aspx</span></p><hr><div style='padding:0 15 3 15;background-color:powderblue'><h2>BBM posted on Friday, April 18, 2008</h2><P>Hi,</P>
<P>I uncovered a situation today while running some test code (NUnit stuff).&nbsp; In its simplest form the problem is...</P>
<P>1)&nbsp; I have a composite BO that contains an Editable Collection of EditableChild BO's.&nbsp; The main BO has a&nbsp;factory "Get"&nbsp;method that returns the whole composite object from a string key.</P>
<P>2)&nbsp;&nbsp;The main BO needs to react to changes in the child objects in the collection.&nbsp;&nbsp;The child objects raise custom events to which the main BO subscribes.</P>
<P>3)&nbsp; The main BO&nbsp;overrides OnDeserialized() to subscribe to the&nbsp;custom child events after the BO is "Fetched".</P>
<P>4)&nbsp;&nbsp;&nbsp;I'm testing the case where the key passed in the Shared Get method does not exist in the database.&nbsp; The DataPortal_Fetch override method in the BO throws an exception when it realizes the database query doesn't return anything.&nbsp; It looks like this...</P><FONT color=#0000ff size=2>
<P>Using</FONT><FONT size=2> dr </FONT><FONT color=#0000ff size=2>As</FONT><FONT size=2> </FONT><FONT color=#0000ff size=2>New</FONT><FONT size=2> SafeDataReader(cm.ExecuteReader)</P>
<P></FONT><FONT color=#0000ff size=2>&nbsp;&nbsp;&nbsp;If</FONT><FONT size=2> dr.Read() </FONT><FONT color=#0000ff size=2>Then</P></FONT><FONT size=2>
<P>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Fetch(dr)</P>
<P></FONT><FONT color=#0000ff size=2>&nbsp;&nbsp;&nbsp;Else</P></FONT><FONT size=2>
<P></FONT><FONT color=#0000ff size=2>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Throw</FONT><FONT size=2> </FONT><FONT color=#0000ff size=2>New</FONT><FONT size=2> System.Data.DataException(</FONT><FONT color=#a31515 size=2>"Could not retrieve an Item for "</FONT><FONT size=2> &amp; Id.ToString)</P>
<P></FONT><FONT color=#0000ff size=2>&nbsp;&nbsp;&nbsp;End</FONT><FONT size=2> </FONT><FONT color=#0000ff size=2>If</P></FONT><FONT size=2>
<P></FONT><FONT color=#0000ff size=2>End</FONT><FONT size=2> </FONT><FONT color=#0000ff size=2>Using</P></FONT>
<P>5)&nbsp; Once the exception is thrown, I see CSLA catch my exception and wrap it in a CallMethodException and then wraps that in a DataPortalException.&nbsp; I'm expecting the DataPortal to return Nothing at this point and the DataPortalException to be caught by my client code.</P>
<P>6)&nbsp; However, even though the exception was thrown in middle of the Fetch process, my OnDeserialized handler is called.&nbsp; That routine attempts to connect the main BO to the custom events in each child of the collection.&nbsp; Since Fetch() was interrupted, the collection is not populated and the attempt to access it causes the system to throw a NullReferenceException.&nbsp; That exception is what gets caught in my NUnit test.&nbsp; I don't know what happens to the DataPortalException.</P>
<P>I guess my question is why,&nbsp;even though&nbsp;Fetch() was interrupted, does the DataPortal attempt to pass back the incomplete BO and trigger the OnDeserialized handler?</P>
<P>I'm using CSLA 3.03 and a "local" (same machine) Remote Dataportal (using .Net Remoting).&nbsp; </P>
<P>Thanks for your help.</P>
<P>BBM</P></div><div style='padding:0 15 3 15;background-color:powderblue'><h2>RockfordLhotka replied on Sunday, April 20, 2008</h2>This is part of the behavior of the DataPortalException - it brings back the object as it was at the time of the exception. The purpose behind doing this is to provide you with complete information for debugging or otherwise attempting to understand the reason for the exception.</div><p style='font-size:small'>Copyright (c) Marimer LLC</body></html>
